src/libs/data-utils.js

```diff
@@ -184,9 +191,101 @@ export const EntityDeleter = ajaxCaller(class EntityDeleter extends Component {
       }, moreToDelete ? `${entity.entityName} (${entity.entityType})` : entity),
       Utils.toIndexPairs(moreToDelete ? additionalDeletions : selectedEntities)),
       div({
-        style: { ...warningStyle, textAlign: 'right' }
+        style: { ...fullWidthWarning, textAlign: 'right' }
       }, [`${selectedEntities.length + additionalDeletions.length} data entries to be deleted.`]),
       deleting && spinnerOverlay
     ])
   }
 })
+
+export const EntityUploader = ajaxCaller(class EntityUploader extends Component {
+  static propTypes = {
+    onDismiss: PropTypes.func.isRequired,
+    onSuccess: PropTypes.func.isRequired,
+    namespace: PropTypes.string.isRequired,
+    name: PropTypes.string.isRequired,
+    entityTypes: PropTypes.array.isRequired
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state = { newEntityType: '' }
+
+    this.uploader = createRef()
+  }
+
+  async doUpload() {
+    const { onDismiss, onSuccess, namespace, name, ajax: { Workspaces } } = this.props
+    const { file } = this.state
+
+    this.setState({ uploading: true })
+
+    try {
+      await Workspaces.workspace(namespace, name).importEntitiesFile(file)
+      onSuccess()
+    } catch (error) {
+      reportError('Error uploading entities', error)
+      onDismiss()
+    }
+  }
+
+  render() {
+    const { onDismiss, entityTypes } = this.props
+    const { uploading, file, newEntityType, isInvalidFile, isInvalidTSV } = this.state
+
+    const inputLabel = text => div({ style: { fontSize: 16, marginBottom: '0.3rem' } }, [text])
+
+    return h(Dropzone, {
+      accept: '.tsv',
+      disableClick: true,
+      style: { flexGrow: 1 },
+      activeStyle: { backgroundColor: colors.blue[3], cursor: 'copy' },
+      ref: this.uploader,
+      onDropRejected: () => this.setState({ file: undefined, isInvalidFile: true, isInvalidTSV: false }),
```

Comment:
nit: since these error states are mutually exclusive, they could be modeled with a single nullable 'enum' string.